lib/scanner/xml-scanner/xml-scanner.c

```diff
@@ -0,0 +1,300 @@
+  if (error && *error)
+    return;
+  g_markup_parse_context_end_parse(self->xml_ctx, error);
+}
+
+
+/*
+  For some patterns, GPatternSpec stores the pattern as reversed
+  string. In such cases, the matching must be executed against the
+  reversed matchstring. If one uses g_pattern_match_string, glib will
+  reverse the string internally in these cases, but it is suboptimal if
+  the same matchstring must be matched against different patterns,
+  because memory is allocated each time and string is copied as
+  reversed, though it would be enough to execute this reverse once. For
+  that reason one can use g_pattern_match(), which accepts both the
+  matchstring and a reversed matchstring as parameters.
+
+  Though there are cases when no reverse is needed at all. This is
+  decided in the constructor of GPatternSpec, but at this time of
+  writing this information is not exported, and cannot be extracted in
+  any safe way, because GPatternSpec is forward declared.
+
+  This function below is an oversimplified/safe version of the logic
+  used glib to decide if the pattern will be reversed or not. One need
+  to note the worst case scenario is if syslog-ng will not reverse the
+  matchstring, though it should because in that case number of extra
+  memory allocations and string reversals scale linearly with the number
+  of patterns. We need to avoid not pre-reversing, when glib would.
+ */
+gboolean
+joker_or_wildcard(GList *patterns)
+{
+  gboolean retval = FALSE;
+  GList *pattern = patterns;
+  while (pattern != NULL)
+    {
+      gchar *str = ((gchar *)pattern->data);
+      if (strpbrk(str, "*?"))
+        {
+          retval = TRUE;
+          break;
+        }
+      pattern = g_list_next(pattern);
+    }
+
+  return retval;
+}
+
+static void
+_compile_and_add(gpointer tag_glob, gpointer exclude_patterns)
+{
+  g_ptr_array_add(exclude_patterns, g_pattern_spec_new(tag_glob));
+}
+
+void
+xml_scanner_init_instance(XMLScanner *self)
+{
+  self->start_element_function = xml_scanner_start_element_method;
+  self->end_element_function = xml_scanner_end_element_method;
+  self->text_function = xml_scanner_text_method;
+  self->key = scratch_buffers_alloc();
+  self->text = scratch_buffers_alloc();
+  self->exclude_patterns = g_ptr_array_new_with_free_func((GDestroyNotify)g_pattern_spec_free);
+}
+
+XMLScanner *
+xml_scanner_new(void)
+{
+  XMLScanner *self = g_new0(XMLScanner, 1);
+  xml_scanner_init_instance(self);
+  return self;
+}
+
+static void
+xml_scanner_init_context(XMLScanner *self)
+{
+  self->scanner_callbacks.start_element = self->start_element_function;
+  self->scanner_callbacks.end_element = self->end_element_function;
+  self->scanner_callbacks.text = self->text_function;
+  self->xml_ctx = g_markup_parse_context_new(&self->scanner_callbacks, 0, self, NULL);
+}
+
+void
+xml_scanner_init(XMLScanner *self, XMLScannerOptions *options, PushCurrentKeyValue push_function,
+                 gpointer user_data, gchar *key_initial)
```

Comment:
Please rename key_initial to key_prefix.

(If you want to move the prefix back to the parser side, there will be a problem: you'll need 2 user data for the push function: the parser instance and the message as well; and you can't cache the message in the parser object, because it is used by multiple threads).

Reply:
Thanks for the additional description about consequences of handling the prefix on the parser side.
Indeed, the push-parser behaviour has some difficulties.

Reply:
És mire lettek figyelmesek?